### PR TITLE
Sum of list 

### LINKED ISSUE #16: Challenge_Easy_Sum_of_List
The function take a list of numbers and return the sum of them.

Input: 
List of numbers 

Output: 
Sum of the list (int)

Sample input: 
[1,2,3,4,5] 

Output: 
15

### PR DESCRIPTION
**Pull request by Emanfalouji**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
  - _It's ok to have extra helper functions if necessary, like with mergesort_
- [x] The function has correct type annotations
- [x] The function is not called at the top level of the function file 
  - _Recursive solutions **can** call the function from **inside** the function body_

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in any docstrings or tests
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/sum_of_list.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        >>> sum_of_list([])
+        Please enter valid numbers separated by spaces.
+    """
+    if not isinstance(numbers, list):
+        raise TypeError("Input must be a list of numbers.")
+    if not all(isinstance(num, (int, float)) for num in numbers):
+        raise ValueError("All elements in the list must be int or float.")
+    return sum(numbers)
+
+if __name__ == "__main__":
```
Comment (Kareiman) [inline solutions/sum_of_list.py]: The function should not be called on the function file

File: solutions/sum_of_list.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    try:
+        # Prompt the user to input a list of numbers separated by spaces
+        user_input = input("Enter a list of numbers separated by spaces: ")
+        # Convert the input string into a list of floats/integers
+        numbers = [float(num) for num in user_input.split()]
+        # Calculate the sum of the list and display the result
+        print("The sum of the list is:", sum_of_list(numbers))
```
Comment (Kareiman) [inline solutions/sum_of_list.py]: Print should not be used within the function file
according to the check list:
There are no print statements anywhere

File: solutions/sum_of_list.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
+"""
+This function asks the user to enter numbers and return the sum of the numbers  .
+
+Created on 05 01 2025
+@author: Eman Alfalouji.
+
+"""
+def sum_of_list(numbers):
```
Comment (Kareiman) [inline solutions/sum_of_list.py]: Should add here the type of the variables, something like the following:
def function_name(variable: type) -> result type:

File: solutions/tests/test_sum_of_list.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        Test that the function correctly sums a list of positive integers.
+        """
+        self.assertEqual(sum_of_list([8, 7, 3, 4, 5]), 27)  # Corrected expected value
+
+    def test_negative_numbers(self):
+        """
+        Test that the function correctly sums a list of negative integers.
+        """
+        self.assertEqual(sum_of_list([-1, -4, -7]), -12)
+
```
Comment (Kareiman) [inline solutions/tests/test_sum_of_list.py:29]: maybe define the following two or three tests as edge or boundary tests

File: solutions/tests/test_sum_of_list.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        Test that the function returns 0 for an empty list.
+        """
+        self.assertEqual(sum_of_list([]), 0)
+
+    def test_single_element(self):
+        """
+        Test that the function returns the single element value when the list has only one number.
+        """
+        self.assertEqual(sum_of_list([10]), 10)
+
```
Comment (Kareiman) [inline solutions/tests/test_sum_of_list.py:45]: You could define them as defensive cases

**Comment by Kareiman** [comment]
Hello Eman,
Impressive work! Only small changes needed to pass the whole checklist!
The last commit you made shows the it didn't pass the formatting check, while the earlier ones passed, so it was a bit confusing for me when it came to the checklist

File: solutions/tests/test_sum_of_list.py (hunk 5 above)
Comment (Emanfalouji) [reply solutions/tests/test_sum_of_list.py:45]: done

**Comment by Emanfalouji** [comment]
> Hello Eman, Impressive work! Only small changes needed to pass the whole checklist! The last commit you made shows the it didn't pass the formatting check, while the earlier ones passed, so it was a bit confusing for me when it came to the checklist

Hi I am working on the commits you wrote The last commit I made is Fix Format which I fixed all the formating issues and it passes all ci checks

**Review by Kareiman — APPROVED** [review]
Great work on this implementation! The code and unit tests are well-structured and thoughtfully written.
Keep up the excellent work @Emanfalouji

**Comment by Emanfalouji** [comment]
Thanks Kareiman your commits was amazing and helpful
